File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 프로그래머스 1844 게임 맵 최단거리 LV.2
from collections import deque


def solution(maps):
    answer = -1
    visited = [[0 for _ in range(len(maps[-1]))] for _ in range(len(maps))]
    op = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    visited[0][0] = 1

    def bfs(start):
        q = deque()
        q.append(start)
        while q:
            x, y = q.popleft()
            for i in range(4):
                nx, ny = x + op[i][0], y + op[i][1]
                if 0 <= nx < len(maps) and 0 <= ny < len(maps[-1]):
                    if visited[nx][ny] == 0 and maps[nx][ny] != 0:
                        q.append((nx, ny))
                        visited[nx][ny] = visited[x][y] + 1

    bfs((0, 0))

    if visited[-1][-1] == 0:
        return answer

    return visited[-1][-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int[] dy = {-1, 0, 1, 0}, dx = {0, 1, 0, -1};
    
    public int bfs(int[][] maps) {
        int w = maps[0].length, h = maps.length;
        boolean[][] visited = new boolean[h][w];
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0, 1});
        visited[0][0] = true;
        
        while (!q.isEmpty()) {
            int[] coord = q.poll();
            int y = coord[0], x = coord[1], turn = coord[2];
            
            if (y == h - 1 && x == w - 1) return turn;
            
            for (int i = 0; i < 4; i++) {
                int ny = y + dy[i], nx = x + dx[i];
                
                if (0 <= ny && ny < h && 0 <= nx && nx < w && !visited[ny][nx] && maps[ny][nx] == 1) {
                    q.offer(new int[]{ny, nx, turn + 1});
                    visited[ny][nx] = true;
                }
            }
        }
        
        return -1;
    }
    
    public int solution(int[][] maps) {
        return bfs(maps);
    }
}